File: plugins/DEV-ESSENTIALS/hooks/console-log-warning.js
#!/usr/bin/env node
/**
 * PostToolUse Hook: Console Log Warning
 *
 * Warns about debug statements (console.log, print(), fmt.Println())
 * in edited files after an Edit operation.
 *
 * Non-blocking (exit code 0).
 */

const MAX_STDIN = 1024 * 1024;

const DEBUG_PATTERNS = [
  { pattern: /console\.log\(/, lang: 'JavaScript/TypeScript' },
  { pattern: /console\.debug\(/, lang: 'JavaScript/TypeScript' },
  { pattern: /\bprint\s*\(/, lang: 'Python' },
  { pattern: /fmt\.Print(ln|f)?\(/, lang: 'Go' },
  { pattern: /System\.out\.print/, lang: 'Java' },
];

let data = '';
process.stdin.setEncoding('utf8');

process.stdin.on('data', chunk => {
  if (data.length < MAX_STDIN) data += chunk.substring(0, MAX_STDIN - data.length);
});

process.stdin.on('end', () => {
  try {
    const disabled = (process.env.DISABLED_HOOKS || '').split(',');
    if (disabled.includes('console-log-warning')) {
      process.stdout.write(data);
      return;
    }

    const input = JSON.parse(data);
    const newString = input.tool_input?.new_string || '';
    const filePath = input.tool_input?.file_path || '';

    if (!newString || /\.(test|spec|_test)\./i.test(filePath)) {
      process.stdout.write(data);
      return;
    }

    const found = [];
    for (const { pattern, lang } of DEBUG_PATTERNS) {
      if (pattern.test(newString)) {
        found.push(lang);
      }
    }

    if (found.length > 0) {
      const langs = [...new Set(found)].join(', ');
      console.error(`[Hook] WARNING: Debug statement detected in ${filePath} (${langs})`);
      console.error('[Hook] Remember to remove debug statements before committing.');
    }

    process.stdout.write(data);
  } catch (err) {
    console.error(`[Hook] console-log-warning error: ${err.message}`);
    process.stdout.write(data);
  }
});

File: plugins/DEV-ESSENTIALS/hooks/console-log-warning.py
#!/usr/bin/env python3
"""
PostToolUse Hook: Console Log Warning

Warns about debug statements (console.log, print(), fmt.Println())
in edited files after an Edit operation.

Non-blocking (exit code 0).
"""

import json
import os
import re
import sys

MAX_STDIN = 1024 * 1024

DEBUG_PATTERNS = [
    (re.compile(r"console\.log\("), "JavaScript/TypeScript"),
    (re.compile(r"console\.debug\("), "JavaScript/TypeScript"),
    (re.compile(r"\bprint\s*\("), "Python"),
    (re.compile(r"fmt\.Print(?:ln|f)?\("), "Go"),
    (re.compile(r"System\.out\.print"), "Java"),
]


def main():
    data = sys.stdin.read(MAX_STDIN)

    try:
        disabled = os.environ.get("DISABLED_HOOKS", "").split(",")
        if "console-log-warning" in disabled:
            sys.stdout.write(data)
            return

        inp = json.loads(data)
        new_string = inp.get("tool_input", {}).get("new_string", "")
        file_path = inp.get("tool_input", {}).get("file_path", "")

        if not new_string or re.search(r"\.(test|spec|_test)\.", file_path, re.IGNORECASE):
            sys.stdout.write(data)
            return

        found = set()
        for pattern, lang in DEBUG_PATTERNS:
            if pattern.search(new_string):
                found.add(lang)

        if found:
            langs = ", ".join(sorted(found))
            print(f"[Hook] WARNING: Debug statement detected in {file_path} ({langs})", file=sys.stderr)
            print("[Hook] Remember to remove debug statements before committing.", file=sys.stderr)

        sys.stdout.write(data)
    except Exception as err:
        print(f"[Hook] console-log-warning error: {err}", file=sys.stderr)
        sys.stdout.write(data)


if __name__ == "__main__":
    main()
